### PR TITLE
Added LLVM ops and lowering phases from standard dialect.

### PR DESCRIPTION
**Pull request by dfki-jugr**
Added LLVM ops and lowering phases from standard dialect for FAbs, FCeil, Cos, FNeg, CopySign and Tanh.

Added test cases for the newly added LLVM operations and lowering features.

### CODE REVIEW

File: include/mlir/Dialect/LLVMIR/LLVMOps.td [hunk 1]
```diff
@@ -684,4 +684,114 @@ def LLVM_ExpOp : LLVM_Op<"intr.exp", [NoSideEffect]>,
...
+                   Results<(outs LLVM_Type:$res)> {
+  let llvmBuilder = [{
+    llvm::Module *module = builder.GetInsertBlock()->getModule();
+    llvm::Function *fn = llvm::Intrinsic::getDeclaration(
+        module, llvm::Intrinsic::copysign, {$a->getType(), $b->getType()});
+    $res = builder.CreateCall(fn, {$a, $b});
+  }];
+}
+
+def LLVM_TanhOp : LLVM_Op<"tanh", [NoSideEffect]>,
```
Comment (ftynse) [inline include/mlir/Dialect/LLVMIR/LLVMOps.td]: If there is no `tanh` op in LLVM, it should not be an LLVM Op. Create it in some dialect and implement a lowering from that dialect to the LLVM dialect instead, so the translation from the LLVM dialect to LLVM IR is as simple as possible. Having magic constants in the translation wouldn't cut it.

If you implement your non-trivial conversion in as a conversion pattern, it is significantly easier to replace.

**Comment by ftynse** [comment]
**please do not merge this** until the concerns on translation are addressed.

File: include/mlir/Dialect/LLVMIR/LLVMOps.td (hunk 1 above)
[the remark this answers is not in the document]
Comment (ftynse) [reply include/mlir/Dialect/LLVMIR/LLVMOps.td]: We need a more detailed explanation of where these values come from and what is computed in the end.

File: include/mlir/Dialect/LLVMIR/LLVMOps.td [hunk 2]
```diff
@@ -684,4 +684,114 @@ def LLVM_ExpOp : LLVM_Op<"intr.exp", [NoSideEffect]>,
...
+                   Arguments<(ins LLVM_Type:$in)>,
+                   Results<(outs LLVM_Type:$res)> {
+  let llvmBuilder = [{
```
[the remark this answers is not in the document]
Comment (pifon2a) [reply include/mlir/Dialect/LLVMIR/LLVMOps.td]: @dfki-jugr Let's remove Tanh from the current PR. This logic be added to legalization from STD TanhOp to LLVM dialect, not from LLVM dialect to LLVM IR.

File: test/Target/llvmir-intrinsics.mlir [hunk 3]
```diff
@@ -18,8 +18,50 @@ llvm.func @exp_test(%arg0: !llvm.float, %arg1: !llvm<"<8 x float>">) {
...
+  "llvm.intr.cos"(%arg1) : (!llvm<"<8 x float>">) -> !llvm<"<8 x float>">
+  llvm.return
+}
+
+// CHECK-LABEL: @copysign_test
+llvm.func @copysign_test(%arg0: !llvm.float, %arg1: !llvm.float, %arg2: !llvm<"<8 x float>">, %arg3: !llvm<"<8 x float>">) {
+  // CHECK: call float @llvm.copysign.f32
+  "llvm.intr.copysign"(%arg0, %arg1) : (!llvm.float, !llvm.float) -> !llvm.float
+  // CHECK: call <8 x float> @llvm.copysign.v8f32
+  "llvm.intr.copysign"(%arg2, %arg3) : (!llvm<"<8 x float>">, !llvm<"<8 x float>">) -> !llvm<"<8 x float>">
```
Comment (pifon2a) [inline test/Target/llvmir-intrinsics.mlir]: tests for `Neg`?

File: include/mlir/Dialect/LLVMIR/LLVMOps.td [hunk 4]
```diff
@@ -662,26 +662,40 @@ def LLVM_ConstantOp
...
+      "{$a->getType(), $b->getType(), $c->getType()}), {$a, $b, $c});"> {
+}
+
+def LLVM_ExpOp : LLVM_UnaryIntrinsicOp<"exp">;
+def LLVM_FAbsOp : LLVM_UnaryIntrinsicOp<"fabs">;
+def LLVM_FCeilOp : LLVM_UnaryIntrinsicOp<"ceil">;
+def LLVM_CosOp : LLVM_UnaryIntrinsicOp<"cos">;
+
+def LLVM_CopySignOp : LLVM_BinaryIntrinsicOp<"copysign">;
+
```
Comment (pifon2a) [inline include/mlir/Dialect/LLVMIR/LLVMOps.td]: nit: remove empty line

File: test/Target/llvmir-intrinsics.mlir (hunk 3 above)
Comment (dfki-jugr) [reply test/Target/llvmir-intrinsics.mlir]: `Neg` is not an intrinsic function in LLVM. The test is already here: https://github.com/tensorflow/mlir/blob/8e269ccd0f6f3e73f4d3fffbd93786836882db97/test/Target/llvmir.mlir#L842

**Pull request closed without merging**